# Assignment 18 : Artificial Neural Networks

## Objective :

Develop a Classification model using Artificial Neural Networks (ANN ) to classify the data points from the "Alphabets_data.csv" dataset into predefined categories of alphabets.

## Task 1 : Data Exploration and Preprocessing

In [1]:
# Import necessary libraries.
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

In [2]:
# Load the dataset.
data = pd.read_csv("Alphabets_data.csv")

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   letter  20000 non-null  object
 1   xbox    20000 non-null  int64 
 2   ybox    20000 non-null  int64 
 3   width   20000 non-null  int64 
 4   height  20000 non-null  int64 
 5   onpix   20000 non-null  int64 
 6   xbar    20000 non-null  int64 
 7   ybar    20000 non-null  int64 
 8   x2bar   20000 non-null  int64 
 9   y2bar   20000 non-null  int64 
 10  xybar   20000 non-null  int64 
 11  x2ybar  20000 non-null  int64 
 12  xy2bar  20000 non-null  int64 
 13  xedge   20000 non-null  int64 
 14  xedgey  20000 non-null  int64 
 15  yedge   20000 non-null  int64 
 16  yedgex  20000 non-null  int64 
dtypes: int64(16), object(1)
memory usage: 2.6+ MB


In [4]:
data.head()

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
0      T     2     8      3       5      1     8    13      0      6      6   
1      I     5    12      3       7      2    10     5      5      4     13   
2      D     4    11      6       8      6    10     6      2      6     10   
3      N     7    11      6       6      3     5     9      4      6      4   
4      G     2     1      3       1      1     8     6      6      6      6   

   x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0      10       8      0       8      0       8  
1       3       9      2       8      4      10  
2       3       7      3       7      3       9  
3       4      10      6      10      2       8  
4       5       9      1       7      5      10

In [5]:
data.letter.unique()

array(['T', 'I', 'D', 'N', 'G', 'S', 'B', 'A', 'J', 'M', 'X', 'O', 'R',
       'F', 'C', 'H', 'W', 'L', 'P', 'E', 'V', 'Y', 'Q', 'U', 'K', 'Z'],
      dtype=object)

In [6]:
data.shape

(20000, 17)

In [7]:
data.size

340000

In [9]:
data.describe()

xbox          ybox         width       height         onpix  \
count  20000.000000  20000.000000  20000.000000  20000.00000  20000.000000   
mean       4.023550      7.035500      5.121850      5.37245      3.505850   
std        1.913212      3.304555      2.014573      2.26139      2.190458   
min        0.000000      0.000000      0.000000      0.00000      0.000000   
25%        3.000000      5.000000      4.000000      4.00000      2.000000   
50%        4.000000      7.000000      5.000000      6.00000      3.000000   
75%        5.000000      9.000000      6.000000      7.00000      5.000000   
max       15.000000     15.000000     15.000000     15.00000     15.000000   

               xbar          ybar         x2bar         y2bar         xybar  \
count  20000.000000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       6.897600      7.500450      4.628600      5.178650      8.282050   
std        2.026035      2.325354      2.699968      2.380823      2.488475   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        6.000000      6.000000      3.000000      4.000000      7.000000   
50%        7.000000      7.000000      4.000000      5.000000      8.000000   
75%        8.000000      9.000000      6.000000      7.000000     10.000000   
max       15.000000     15.000000     15.000000     15.000000     15.000000   

            x2ybar        xy2bar         xedge        xedgey         yedge  \
count  20000.00000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       6.45400      7.929000      3.046100      8.338850      3.691750   
std        2.63107      2.080619      2.332541      1.546722      2.567073   
min        0.00000      0.000000      0.000000      0.000000      0.000000   
25%        5.00000      7.000000      1.000000      8.000000      2.000000   
50%        6.00000      8.000000      3.000000      8.000000      3.000000   
75%        8.00000      9.000000      4.000000      9.000000      5.000000   
max       15.00000     15.000000     15.000000     15.000000     15.000000   

            yedgex  
count  20000.00000  
mean       7.80120  
std        1.61747  
min        0.00000  
25%        7.00000  
50%        8.00000  
75%        9.00000  
max       15.00000

In [10]:
data.duplicated().sum()

1332

In [11]:
# Remove duplicate rows.
data.drop_duplicates(inplace=True)

In [12]:
data.duplicated().sum()

0

In [18]:
data.isna().sum().sum()

0

In [13]:
data.columns

Index(['letter', 'xbox', 'ybox', 'width', 'height', 'onpix', 'xbar', 'ybar',
       'x2bar', 'y2bar', 'xybar', 'x2ybar', 'xy2bar', 'xedge', 'xedgey',
       'yedge', 'yedgex'],
      dtype='object')

In [14]:
# Divide the data into X and Y
X = data.drop("letter",axis=1)

In [15]:
Y = data.letter

In [16]:
X.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18668 entries, 0 to 19999
Data columns (total 16 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   xbox    18668 non-null  int64
 1   ybox    18668 non-null  int64
 2   width   18668 non-null  int64
 3   height  18668 non-null  int64
 4   onpix   18668 non-null  int64
 5   xbar    18668 non-null  int64
 6   ybar    18668 non-null  int64
 7   x2bar   18668 non-null  int64
 8   y2bar   18668 non-null  int64
 9   xybar   18668 non-null  int64
 10  x2ybar  18668 non-null  int64
 11  xy2bar  18668 non-null  int64
 12  xedge   18668 non-null  int64
 13  xedgey  18668 non-null  int64
 14  yedge   18668 non-null  int64
 15  yedgex  18668 non-null  int64
dtypes: int64(16)
memory usage: 2.4 MB


In [17]:
Y.info()

<class 'pandas.core.series.Series'>
Int64Index: 18668 entries, 0 to 19999
Series name: letter
Non-Null Count  Dtype 
--------------  ----- 
18668 non-null  object
dtypes: object(1)
memory usage: 291.7+ KB


In [19]:
# Label encoding the Y classes.
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
Y_labels = encoder.fit_transform(Y)

In [20]:
Y_labels

array([19,  8,  3, ...,  2, 19,  0])

In [22]:
!pip install tensorflow

     ---------------------------------------- 0.0/48.7 kB ? eta -:--:--
     ------------------------------- ------ 41.0/48.7 kB 991.0 kB/s eta 0:00:01
     -------------------------------------- 48.7/48.7 kB 620.1 kB/s eta 0:00:00
   ---------------------------------------- 0.0/385.0 MB ? eta -:--:--
   ---------------------------------------- 0.1/385.0 MB 3.3 MB/s eta 0:01:57
   ---------------------------------------- 0.2/385.0 MB 3.3 MB/s eta 0:01:57
   ---------------------------------------- 0.5/385.0 MB 4.2 MB/s eta 0:01:31
   ---------------------------------------- 1.0/385.0 MB 5.7 MB/s eta 0:01:08
   ---------------------------------------- 1.7/385.0 MB 7.8 MB/s eta 0:00:50
   ---------------------------------------- 2.3/385.0 MB 8.7 MB/s eta 0:00:44
   ---------------------------------------- 3.0/385.0 MB 9.5 MB/s eta 0:00:41
   ---------------------------------------- 3.7/385.0 MB 10.4 MB/s eta 0:00:37
   ---------------------------------------- 4.5/385.0 MB 11.1 MB/s eta 0

In [25]:
from tensorflow.keras.utils import to_categorical
Y_labels = to_categorical(Y_labels)

In [26]:
Y_labels

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 1., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [1., 0., 0., ..., 0., 0., 0.]])

In [28]:
# Normalize the input features.

from sklearn.preprocessing import StandardScaler

In [29]:
scaler = StandardScaler()
scaled_df = scaler.fit_transform(X)
X = pd.DataFrame(scaled_df,columns=X.columns)

In [30]:
X

xbox      ybox     width    height     onpix      xbar      ybar  \
0     -1.144417  0.237070 -1.178294 -0.232716 -1.231474  0.531756  2.358246   
1      0.451855  1.491943 -1.178294  0.685459 -0.768617  1.515924 -1.090370   
2     -0.080235  1.178224  0.372167  1.144546  1.082808  1.515924 -0.659293   
3      1.516037  1.178224  0.372167  0.226371 -0.305761 -0.944497  0.633938   
4     -1.144417 -1.958957 -1.178294 -2.069066 -1.231474  0.531756 -0.659293   
...         ...       ...       ...       ...       ...       ...       ...   
18663  0.451855  0.237070  0.888987  0.685459  1.545665  0.039671  0.633938   
18664 -1.144417 -1.645239 -1.178294 -1.150891 -0.768617  0.039671 -0.228216   
18665  1.516037  0.864506  1.405808  1.144546  0.157095 -1.436582  0.202861   
18666  0.983946  0.550788  0.372167  0.685459  0.619952 -0.452413  1.496092   
18667 -0.080235  0.550788  0.372167  0.226371 -0.768617  1.023840 -1.090370   

          x2bar     y2bar     xybar    x2ybar    xy2bar     xedge    xedgey  \
0     -1.746100  0.344235 -0.953149  1.351557  0.035798 -1.347535 -0.223412   
1      0.158195 -0.517819  1.882192 -1.315513  0.510429 -0.486890 -0.223412   
2     -0.984382  0.344235  0.667046 -1.315513 -0.438833 -0.056568 -0.858409   
3     -0.222664  0.344235 -1.763246 -0.934503  0.985060  1.234399  1.046582   
4      0.539054  0.344235 -0.953149 -0.553493  0.510429 -0.917213 -0.858409   
...         ...       ...       ...       ...       ...       ...       ...   
18663 -0.222664  1.206289 -0.548100  0.208527  0.035798 -0.056568  1.046582   
18664  0.919913  0.344235 -0.953149 -0.172483 -1.862725 -0.486890 -0.223412   
18665  0.539054  1.637317  1.477143  0.970547  2.408953 -0.486890  0.411585   
18666 -0.603523  0.775262  1.072095  0.970547 -1.388094 -0.486890  2.316576   
18667 -0.603523 -1.810900 -0.143052 -2.077534  0.035798 -0.486890 -0.858409   

          yedge    yedgex  
0     -1.494712  0.130096  
1      0.079965  1.336868  
2     -0.313704  0.733482  
3     -0.707373  0.130096  
4      0.473635  1.336868  
...         ...       ...  
18663  1.654643 -1.076676  
18664 -0.313704 -0.473290  
18665 -0.313704 -0.473290  
18666 -0.707373 -2.283449  
18667 -0.707373  0.130096  

[18668 rows x 16 columns]

In [31]:
X.describe()

xbox          ybox         width        height         onpix  \
count  1.866800e+04  1.866800e+04  1.866800e+04  1.866800e+04  1.866800e+04   
mean   1.583382e-16  1.035288e-16  1.826979e-17  1.796530e-16  9.896138e-17   
std    1.000027e+00  1.000027e+00  1.000027e+00  1.000027e+00  1.000027e+00   
min   -2.208599e+00 -2.272675e+00 -2.728755e+00 -2.528154e+00 -1.694330e+00   
25%   -6.123265e-01 -7.040845e-01 -6.614736e-01 -6.918037e-01 -7.686173e-01   
50%   -8.023549e-02  2.370700e-01 -1.446532e-01  2.263714e-01 -3.057609e-01   
75%    4.518555e-01  8.645063e-01  3.721671e-01  6.854589e-01  6.199519e-01   
max    5.772765e+00  2.433097e+00  5.023550e+00  4.358159e+00  5.248516e+00   

               xbar          ybar         x2bar         y2bar         xybar  \
count  1.866800e+04  1.866800e+04  1.866800e+04  1.866800e+04  1.866800e+04   
mean   1.400684e-16 -8.563966e-17  1.583382e-16  9.667766e-17 -2.890814e-16   
std    1.000027e+00  1.000027e+00  1.000027e+00  1.000027e+00  1.000027e+00   
min   -3.404919e+00 -3.245754e+00 -1.746100e+00 -2.241927e+00 -3.383441e+00   
25%   -4.524129e-01 -6.592928e-01 -6.035232e-01 -5.178190e-01 -5.481003e-01   
50%    3.967147e-02 -2.282158e-01 -2.226642e-01 -8.679188e-02 -1.430515e-01   
75%    5.317558e-01  6.339380e-01  5.390538e-01  7.752623e-01  6.670459e-01   
max    3.976346e+00  3.220400e+00  3.966785e+00  4.223479e+00  2.692289e+00   

             x2ybar        xy2bar         xedge        xedgey         yedge  \
count  1.866800e+04  1.866800e+04  1.866800e+04  1.866800e+04  1.866800e+04   
mean   1.484421e-16 -6.813111e-17  6.280242e-17 -7.993035e-18  1.979228e-17   
std    1.000027e+00  1.000027e+00  1.000027e+00  1.000027e+00  1.000027e+00   
min   -2.458544e+00 -3.761249e+00 -1.347535e+00 -5.303388e+00 -1.494712e+00   
25%   -5.534933e-01 -4.388327e-01 -4.868905e-01 -2.234123e-01 -7.073732e-01   
50%   -1.724832e-01  3.579817e-02 -5.656800e-02 -2.234123e-01  7.996539e-02   
75%    5.895370e-01  5.104290e-01  3.737545e-01  4.115847e-01  4.736347e-01   
max    3.256607e+00  3.358214e+00  5.107302e+00  4.221567e+00  4.410328e+00   

             yedgex  
count  1.866800e+04  
mean   2.055352e-17  
std    1.000027e+00  
min   -4.696993e+00  
25%   -4.732903e-01  
50%    1.300958e-01  
75%    7.334820e-01  
max    4.353799e+00

## Task 2 : Model Implementation

In [32]:
# Split the data into train and test.
from sklearn.model_selection import train_test_split

train_x,test_x,train_y,test_y = train_test_split(X,Y_labels,test_size = 0.2 , random_state = 7)

In [33]:
print("train_x",train_x.shape)
print("train_y",train_y.shape)
print("test_x",test_x.shape)
print("test_y",test_y.shape)

train_x (14934, 16)
train_y (14934, 26)
test_x (3734, 16)
test_y (3734, 26)


In [34]:
# Model building 
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense


In [35]:
def create_model(optimizer="adam",activation="relu"):
    model = Sequential()
    model.add(Dense(64,input_dim=train_x.shape[1],activation=activation))
    model.add(Dense(32,activation=activation))
    model.add(Dense(Y_labels.shape[1],activation="softmax"))
    model.compile(loss="categorical_crossentropy",optimizer=optimizer ,metrics = ['accuracy'])
    return model

In [36]:
model = create_model()

In [37]:
model.fit(train_x,train_y,epochs=12,batch_size=30,verbose=1)

Epoch 1/12
498/498 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.2945 - loss: 2.5477
Epoch 2/12
498/498 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7222 - loss: 0.9950
Epoch 3/12
498/498 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7821 - loss: 0.7561
Epoch 4/12
498/498 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8109 - loss: 0.6328
Epoch 5/12
498/498 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8421 - loss: 0.5488
Epoch 6/12
498/498 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8565 - loss: 0.4927
Epoch 7/12
498/498 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8765 - loss: 0.4387
Epoch 8/12
498/498 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8796 - loss: 0.4120
Epoch 9/12
498/498 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8876 - loss: 0.3847
Epoch 10/12
498/498 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8969 - loss: 0.3449
Epoch 11/12
498/498 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8983 - loss: 0.3282
Epoch 12/12
498/498 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step

In [38]:
# Evaluation
y_pred = model.predict(test_x)

117/117 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step


In [39]:
y_pred

array([[2.6259952e-09, 7.5594103e-01, 7.5722448e-07, ..., 1.2069548e-06,
        7.9300638e-10, 5.1497604e-04],
       [4.4571520e-06, 4.4710975e-04, 9.6155535e-03, ..., 2.0341472e-06,
        1.3741787e-06, 7.8664216e-06],
       [9.9612385e-01, 1.3508266e-14, 1.7291725e-08, ..., 3.2118950e-03,
        1.1289103e-13, 1.3686055e-09],
       ...,
       [3.1483033e-08, 9.9257565e-01, 7.2699304e-09, ..., 1.4491045e-06,
        4.5238761e-07, 2.3004414e-08],
       [8.1310652e-08, 6.8901485e-05, 2.0192349e-05, ..., 5.3470912e-03,
        1.7632800e-07, 2.3797782e-01],
       [3.3094556e-04, 1.3920305e-04, 4.1970685e-05, ..., 1.3612978e-02,
        4.0086943e-06, 1.5003145e-04]], dtype=float32)

In [41]:
y_pred_classes =np.argmax(y_pred,axis=1)
y_pred_classes

array([ 1, 14,  0, ...,  1, 18,  6], dtype=int64)

In [43]:
y_classes = np.argmax(test_y,axis=1)

In [44]:
y_classes

array([18,  7,  0, ...,  1, 18,  6], dtype=int64)

In [45]:
from sklearn.metrics import classification_report,accuracy_score
print(classification_report(y_classes,y_pred_classes))

              precision    recall  f1-score   support

           0       0.97      0.94      0.96       151
           1       0.78      0.94      0.85       147
           2       0.94      0.90      0.92       142
           3       0.91      0.90      0.91       159
           4       0.87      0.83      0.85       141
           5       0.92      0.89      0.90       148
           6       0.84      0.84      0.84       135
           7       0.75      0.75      0.75       148
           8       0.93      0.89      0.91       111
           9       0.95      0.91      0.93       141
          10       0.83      0.92      0.87       146
          11       0.90      0.85      0.88       129
          12       0.92      0.91      0.92       120
          13       0.93      0.91      0.92       128
          14       0.88      0.92      0.90       155
          15       0.93      0.93      0.93       146
          16       0.92      0.92      0.92       153
          17       0.90    

In [48]:
print(accuracy_score(y_classes,y_pred_classes)*100)

89.55543652919121


In [49]:
X.shape

(18668, 16)

In [51]:
!pip install keras

In [55]:
!pip install scikeras


## Task 3 : Hyperparameter Tuning

In [66]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from scikeras.wrappers import KerasClassifier

def build_model(optimizer='adam', **kwargs):
    model = Sequential()
    model.add(Dense(units=64, input_shape=(16,), activation='relu'))
    model.add(Dense(units=32, activation='relu'))
    model.add(Dense(units=26, activation='softmax'))
    model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])
    return model


In [70]:

model = KerasClassifier(model=build_model, epochs=10, batch_size=32)


In [71]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'model__optimizer': ['adam', 'rmsprop'],
    'epochs': [10, 20],
    'batch_size': [32, 64]
}

grid_model = GridSearchCV(estimator=model, param_grid=param_grid, cv=3)

# Fit the model
grid = grid_model.fit(train_x, train_y)




Epoch 1/10
312/312 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.2501 - loss: 2.7616
Epoch 2/10
312/312 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6724 - loss: 1.1823
Epoch 3/10
312/312 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7478 - loss: 0.8929
Epoch 4/10
312/312 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7836 - loss: 0.7617
Epoch 5/10
312/312 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8112 - loss: 0.6585
Epoch 6/10
312/312 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8287 - loss: 0.5841
Epoch 7/10
312/312 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8519 - loss: 0.5243
Epoch 8/10
312/312 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8592 - loss: 0.4873
Epoch 9/10
312/312 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8673 - loss: 0.4502
Epoch 10/10
312/312 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8814 - loss: 0.4095
156/156 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
Epoch 1/10
312/312 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.2544 - loss: 2.7495
Epoch 2/10
3

In [73]:
# Retriving the best parameters from the grid.
tuned_model = grid.best_estimator_
print("Best Parameters: ", grid.best_params_)

Best Parameters:  {'batch_size': 32, 'epochs': 20, 'model__optimizer': 'adam'}


In [74]:
tuned_model

KerasClassifier(
	model=<function build_model at 0x0000029279C8C180>
	build_fn=None
	warm_start=False
	random_state=None
	optimizer=rmsprop
	loss=None
	metrics=None
	batch_size=32
	validation_batch_size=None
	verbose=1
	callbacks=None
	validation_split=0.0
	shuffle=True
	run_eagerly=False
	epochs=20
	class_weight=None
	model__optimizer=adam
)

In [76]:
y_pred = tuned_model.predict(test_x)
y_pred_classes = np.argmax(y_pred,axis=1)

117/117 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step


In [77]:
y_pred_classes

array([ 1,  7,  0, ...,  1, 18,  6], dtype=int64)

In [80]:
print(accuracy_score(y_classes,y_pred_classes)*100)

92.28709159078736


In [81]:
print(classification_report(y_classes,y_pred_classes))

              precision    recall  f1-score   support

           0       0.97      0.97      0.97       151
           1       0.83      0.92      0.87       147
           2       0.89      0.92      0.91       142
           3       0.90      0.96      0.93       159
           4       0.86      0.91      0.88       141
           5       0.95      0.90      0.92       148
           6       0.93      0.85      0.89       135
           7       0.89      0.80      0.84       148
           8       0.96      0.90      0.93       111
           9       0.93      0.96      0.94       141
          10       0.89      0.90      0.90       146
          11       0.98      0.90      0.94       129
          12       0.95      0.95      0.95       120
          13       0.88      0.90      0.89       128
          14       0.92      0.93      0.93       155
          15       0.95      0.94      0.94       146
          16       0.93      0.95      0.94       153
          17       0.90    

# Comparing the Model with default parameters and with tuned parameters.


### Overall Performance

- **Accuracy:**
  - **Before Tuning:** 0.90
  - **After Tuning:** 0.92
  - **Improvement:** The accuracy increased by 2%, indicating an overall improvement in the model's ability to correctly classify instances.

- **Macro Avg (Precision/Recall/F1-Score):**
  - **Before Tuning:** 0.90 (Precision), 0.90 (Recall), 0.90 (F1-Score)
  - **After Tuning:** 0.92 (Precision), 0.92 (Recall), 0.92 (F1-Score)
  - **Improvement:** All metrics show a 2% increase, suggesting that the model is more balanced in its performance across all classes after tuning.

- **Weighted Avg (Precision/Recall/F1-Score):**
  - **Before Tuning:** 0.90 (Precision), 0.90 (Recall), 0.90 (F1-Score)
  - **After Tuning:** 0.92 (Precision), 0.92 (Recall), 0.92 (F1-Score)
  - **Improvement:** Similar to macro averages, the weighted averages improved by 2%, showing better overall performance, particularly in classes with more instances.

### Class-Specific Performance

1. **Class 0:**
   - **Before Tuning:** F1-Score: 0.96
   - **After Tuning:** F1-Score: 0.97
   - **Improvement:** Slight improvement in precision, recall, and F1-score.

2. **Class 1:**
   - **Before Tuning:** F1-Score: 0.85
   - **After Tuning:** F1-Score: 0.87
   - **Improvement:** Significant increase in recall (from 0.94 to 0.92) and a corresponding improvement in the F1-score.

3. **Class 3:**
   - **Before Tuning:** F1-Score: 0.91
   - **After Tuning:** F1-Score: 0.93
   - **Improvement:** Precision and recall both improved, leading to a higher F1-score.

4. **Class 7:**
   - **Before Tuning:** F1-Score: 0.75
   - **After Tuning:** F1-Score: 0.84
   - **Improvement:** Significant improvement across precision, recall, and F1-score, indicating better performance on this class.

5. **Class 11:**
   - **Before Tuning:** F1-Score: 0.88
   - **After Tuning:** F1-Score: 0.94
   - **Improvement:** Substantial improvement, particularly in precision, with a corresponding increase in the F1-score.

6. **Class 25:**
   - **Before Tuning:** F1-Score: 0.93
   - **After Tuning:** F1-Score: 0.94
   - **Improvement:** Slight improvement across all metrics.

### General Observations

- **Consistency:** After hyperparameter tuning, the model shows more consistent performance across all classes. Even classes that had lower scores initially (e.g., Class 1, Class 7) have improved, indicating better overall robustness.
- **Recall Improvement:** Several classes (e.g., Classes 4, 7, 11) saw significant recall improvements, meaning the model is better at identifying instances of these classes.
- **Precision Stability:** Precision either remained stable or improved slightly for most classes, contributing to the overall rise in accuracy.

### Conclusion

The hyperparameter tuning has led to a generally improved model with higher accuracy and more consistent performance across all classes. The gains in recall and precision, especially in previously underperforming classes, indicate that the tuning has effectively enhanced the model's ability to generalize to unseen data.